src/dune_rules/ctypes/ctypes_rules.ml

```diff
@@ -216,33 +216,57 @@ let build_c_program ~foreign_archives_deps ~sctx ~dir ~source_files ~scope
     in
     deps
   in
-  let build =
-    let absolute_path_hack p =
-      (* These normal path builder things construct relative paths like
-         _build/default/your/project/file.c but before dune runs gcc it actually
-         cds into _build/default, which fails, so we turn them into absolutes to
-         hack around it. *)
-      Path.relative (Path.build dir) p |> Path.to_absolute_filename
-    in
-    let action =
-      let open Action_builder.O in
-      let* include_args = Resolve.Memo.read include_args in
-      let* base_args = with_user_and_std_flags in
-      deps
-      >>> Action_builder.map cflags ~f:(fun cflags_args ->
-              let source_files = List.map source_files ~f:absolute_path_hack in
-              let output = absolute_path_hack output in
-              let args =
-                base_args @ cflags_args @ include_args @ source_files
-                @ [ "-o"; output ]
-              in
-              Action.run exe args)
-    in
-    Action_builder.with_file_targets action
-      ~file_targets:[ Path.Build.relative dir output ]
+  let action =
+    if version >= (0, 3) then
+      let of_action_builder l : _ Command.Args.t =
+        Dyn (Action_builder.map l ~f:(fun l -> Command.Args.As l))
+      in
```

Comment:
What about changing `with_user_and_std_flags`, `cflags`, `include_args` to return a `Command.Args.t` and then concatenating everything with `S` for `version < (0, 3)`. We should avoid using `@` when building command line arguments.

Reply:
That's my end goal (using `Command.run` everywhere) but the "old" branch is using a plain `Action.run` so I don't think I can use any `Args.t` over there. I think it's going to be easier to clean up once the old branch is deleted, but if you can think of a way to convert the `Args.t` to a list of strings we can do that.

Reply:
Hmm I'll try again something with `Command.expand`. I think that the targets part was causing problems but it might be ok for just the flags part.

Reply:
OK, that worked!